src/CommandLine.cc

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+
+#include "CommandLine.h"
+
+static void
+ResetGetopt(const bool allowStderrWarnings)
+{
+    opterr = allowStderrWarnings ? 1 : 0;
+    // getopt(3) uses global state but resets it if optind is zero
+    // getopt(3) always skips argv[0], even if optind is zero
+    optind = 0;
+}
+
+CommandLine::CommandLine(int argC, char *argV[], const char *shortRules, const struct option *longRules):
+    argv_(),
+    shortOptions_(xstrdup(shortRules)),
+    longOptions_()
+{
+    assert(argC > 0); // C++ main() requirement that makes our arg0() safe
+
+    /* copy argV items */
+    argv_.reserve(argC+1);
+    for (int i = 0; i < argC; ++i)
+        argv_.push_back(xstrdup(argV[i]));
+    argv_.push_back(nullptr); // POSIX argv "must be terminated by a null pointer"
+
+    /* copy grammar rules for the long options */
+    for (auto longOption = longRules; longOption; ++longOption) {
+        longOptions_.push_back(*longOption);
```

Comment:
BTW, this operation ignores the fact that struct option has pointers (possibly dynamically allocated), so we can end up with 'dangling pointers'.

Reply:
Please fix by adding a class wrapper for `options` that will dupe/free the options name.

Reply:
Another alternative to consider is moving grammar rules from the constructor to the two parsing-related methods so that we do not have to store/clone the rules. Such a move would result in slightly clunkier callers, but perhaps it is still a good idea given the complexities of cloning rules information. 

I actually started my rewrite with that design, but rejected it. My rejection was partially based on how the original main.cc caller looked, but I later realized that main.cc misinterpreted the last getopt_long() parameter. Without that parameter, the callers would not look so bad, and we would not have to worry about cloning.

In the future, somebody can add GramaticalCommandLine or a similar class to add grammar rules if needed.

Reply:
I assume that the wrapper should inherit the _option_ C structure (in order to be passed to getopt_long() instead of _option_).

Reply:
Right. Please note that the alternative design does not need a wrapper.

Reply:
Done with the wrapper.